example/build-math-application-with-agent/build-math-application-with-agent.ipynb

```diff
@@ -0,0 +1,369 @@
+   "execution_count": 16,
+   "outputs": [],
+   "source": [
+    "from promptulate.llms import ChatOpenAI\n",
+    "from promptulate.tools.wikipedia.tools import wikipedia_search\n",
+    "from promptulate.tools.math.tools import calculator\n",
+    "from promptulate.schema import MessageSet, SystemMessage, UserMessage, AssistantMessage\n",
+    "import promptulate as pne"
+   ],
+   "metadata": {
+    "collapsed": false,
+    "ExecuteTime": {
+     "end_time": "2024-04-06T16:11:11.007058Z",
+     "start_time": "2024-04-06T16:11:10.988121100Z"
+    }
+   },
+   "id": "bf24915502504abb"
+  },
+  {
+   "cell_type": "markdown",
+   "source": [
+    "### Step 2\n",
+    "Next, we will define our OpenAI-based Language Model.The architectural design of `promptulate` is easily compatible with different large language model extensions. In `promptulate`, llm is responsible for the most basic part of content generation, so it is the most basic component.By default, `ChatOpenAI` in `promptulate` uses the `gpt-3.5-turbo` model."
+   ],
+   "metadata": {
+    "collapsed": false
+   },
+   "id": "24cf6e82bd6c908e"
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 17,
+   "outputs": [],
+   "source": [
+    "llm = ChatOpenAI()"
+   ],
+   "metadata": {
+    "collapsed": false,
+    "ExecuteTime": {
+     "end_time": "2024-04-06T16:11:11.020869300Z",
+     "start_time": "2024-04-06T16:11:11.007058Z"
+    }
+   },
+   "id": "ca8d4ef8b9b9bcc3"
+  },
+  {
+   "cell_type": "markdown",
+   "source": [
+    "We would be using this LLM both within our math and reasoning process and as the decision maker for our agent."
+   ],
+   "metadata": {
+    "collapsed": false
+   },
+   "id": "479c360af07d9a6"
+  },
+  {
+   "cell_type": "markdown",
+   "source": [
+    "### Step 3\n",
+    "When constructing your own agent, you will need to provide it with a list of tools that it can use. Difine a tool， the only you need to do is to provide a function to Promptulate. Promptulate will automatically convert it to a tool that can be used by the language learning model (LLM). The final presentation result it presents to LLM is an OpenAI type JSON schema declaration.\n",
+    "\n",
+    "Actually, Promptulate will analysis function name, parameters type, parameters attribution, annotations and docs when you provide the function. We strongly recommend that you use the official best practices of Template for function writing. The best implementation of a function requires adding type declarations to its parameters and providing function level annotations. Ideally, declare the meaning of each parameter within the annotations.\n",
+    "\n",
+    "We will now create our three tools. The first one will be the online tool using the Wikipedia API wrapper:"
+   ],
+   "metadata": {
+    "collapsed": false
+   },
+   "id": "a2b9b2c90ec135c6"
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 18,
+   "outputs": [],
+   "source": [
+    "# Wikipedia Tool\n",
```

Comment:
You have import the relevant tool, why you need wrap it instead of using these tools directly?